// number-to-words-ru/dist/index.js
const B = typeof self == "object" ? self : globalThis, K = (e, A) => {
  const N = (E, V) => (e.set(V, E), E), t = (E) => {
    if (e.has(E))
      return e.get(E);
    const [V, r] = A[E];
    switch (V) {
      case 0:
      case -1:
        return N(r, E);
      case 1: {
        const n = N([], E);
        for (const s of r)
          n.push(t(s));
        return n;
      }
      case 2: {
        const n = N({}, E);
        for (const [s, S] of r)
          n[t(s)] = t(S);
        return n;
      }
      case 3:
        return N(new Date(r), E);
      case 4: {
        const { source: n, flags: s } = r;
        return N(new RegExp(n, s), E);
      }
      case 5: {
        const n = N(/* @__PURE__ */ new Map(), E);
        for (const [s, S] of r)
          n.set(t(s), t(S));
        return n;
      }
      case 6: {
        const n = N(/* @__PURE__ */ new Set(), E);
        for (const s of r)
          n.add(t(s));
        return n;
      }
      case 7: {
        const { name: n, message: s } = r;
        return N(new B[n](s), E);
      }
      case 8:
        return N(BigInt(r), E);
      case "BigInt":
        return N(Object(BigInt(r)), E);
    }
    return N(new B[V](r), E);
  };
  return t;
}, $ = (e) => K(/* @__PURE__ */ new Map(), e)(0), U = "", { toString: Q } = {}, { keys: II } = Object, l = (e) => {
  const A = typeof e;
  if (A !== "object" || !e)
    return [0, A];
  const N = Q.call(e).slice(8, -1);
  switch (N) {
    case "Array":
      return [1, U];
    case "Object":
      return [2, U];
    case "Date":
      return [3, U];
    case "RegExp":
      return [4, U];
    case "Map":
      return [5, U];
    case "Set":
      return [6, U];
  }
  return N.includes("Array") ? [1, N] : N.includes("Error") ? [7, N] : [2, N];
}, p = ([e, A]) => e === 0 && (A === "function" || A === "symbol"), EI = (e, A, N, t) => {
  const E = (r, n) => {
    const s = t.push(r) - 1;
    return N.set(n, s), s;
  }, V = (r) => {
    if (N.has(r))
      return N.get(r);
    let [n, s] = l(r);
    switch (n) {
      case 0: {
        let c = r;
        switch (s) {
          case "bigint":
            n = 8, c = r.toString();
            break;
          case "function":
          case "symbol":
            if (e)
              throw new TypeError("unable to serialize " + s);
            c = null;
            break;
          case "undefined":
            return E([-1], r);
        }
        return E([n, c], r);
      }
      case 1: {
        if (s)
          return E([s, [...r]], r);
        const c = [], o = E([n, c], r);
        for (const O of r)
          c.push(V(O));
        return o;
      }
      case 2: {
        if (s)
          switch (s) {
            case "BigInt":
              return E([s, r.toString()], r);
            case "Boolean":
            case "Number":
            case "String":
              return E([s, r.valueOf()], r);
          }
        if (A && "toJSON" in r)
          return V(r.toJSON());
        const c = [], o = E([n, c], r);
        for (const O of II(r))
          (e || !p(l(r[O]))) && c.push([V(O), V(r[O])]);
        return o;
      }
      case 3:
        return E([n, r.toISOString()], r);
      case 4: {
        const { source: c, flags: o } = r;
        return E([n, { source: c, flags: o }], r);
      }
      case 5: {
        const c = [], o = E([n, c], r);
        for (const [O, M] of r)
          (e || !(p(l(O)) || p(l(M)))) && c.push([V(O), V(M)]);
        return o;
      }
      case 6: {
        const c = [], o = E([n, c], r);
        for (const O of r)
          (e || !p(l(O))) && c.push(V(O));
        return o;
      }
    }
    const { message: S } = r;
    return E([n, { name: s, message: S }], r);
  };
  return V;
}, x = (e, { json: A, lossy: N } = {}) => {
  const t = [];
  return EI(!(A || N), !!A, /* @__PURE__ */ new Map(), t)(e), t;
}, TI = typeof structuredClone == "function" ? (
  /* c8 ignore start */
  (e, A) => A && ("json" in A || "lossy" in A) ? $(x(e, A)) : structuredClone(e)
) : (e, A) => $(x(e, A));
function f(e, A, N, t) {
  const V = A[N][e];
  return typeof V == "object" ? V[t] : V;
}
const I = {
  NOMINATIVE: "nominative",
  // именительный
  GENITIVE: "genitive",
  // родительный
  DATIVE: "dative",
  // дательный
  ACCUSATIVE: "accusative",
  // винительный
  INSTRUMENTAL: "instrumental",
  // творительный
  PREPOSITIONAL: "prepositional"
  // предложный
}, j = [
  "миллион",
  "миллиард",
  "триллион",
  "квадриллион",
  "квинтиллион",
  "секстиллион",
  "септиллион",
  "октиллион",
  "нониллион",
  "дециллион",
  "ундециллион",
  "додециллион",
  "тредециллион",
  "кваттуордециллион",
  "квиндециллион",
  "сексдециллион",
  // 10^51
  "септемдециллион",
  "октодециллион",
  "новемдециллион",
  "вигинтиллион",
  "унвигинтиллион",
  "дуовигинтиллион",
  "тревигинтиллион",
  "кватуорвигинтиллион",
  "квинвигинтиллион",
  "сексвигинтиллион",
  "септенвигинтиллион",
  "октовигинтиллион",
  "новемвигинтиллион",
  "тригинтиллион",
  "унтригинтиллион",
  "дуотригинтиллион",
  "третригинтиллион",
  // 10^102
  "кватортригинтиллион",
  "квинтригинтиллион",
  "секстригинтиллион",
  "септентригинтиллион",
  "октотригинтиллион",
  "новемтригинтиллион",
  "квадрагинтиллион",
  "унквадрагинтиллион",
  "дуоквадрагинтиллион",
  "треквадрагинтиллион",
  "кваторквадрагинтиллион",
  "квинквадрагинтиллион",
  "сексквадрагинтиллион",
  "септенквадрагинтиллион",
  "октоквадрагинтиллион",
  "новемквадрагинтиллион",
  // 10^150
  "квинквагинтиллион",
  "унквинкагинтиллион",
  "дуоквинкагинтиллион",
  "треквинкагинтиллион",
  "кваторквинкагинтиллион",
  "квинквинкагинтиллион",
  "сексквинкагинтиллион",
  "септенквинкагинтиллион",
  "октоквинкагинтиллион",
  "новемквинкагинтиллион",
  "сексагинтиллион",
  "унсексагинтиллион",
  "дуосексагинтиллион",
  "тресексагинтиллион",
  "кваторсексагинтиллион",
  "квинсексагинтиллион",
  "секссексагинтиллион",
  // 10^201
  "септенсексагинтиллион",
  "октосексагинтиллион",
  "новемсексагинтиллион",
  "септагинтиллион",
  "унсептагинтиллион",
  "дуосептагинтиллион",
  "тресептагинтиллион",
  "кваторсептагинтиллион",
  "квинсептагинтиллион",
  "секссептагинтиллион",
  "септенсептагинтиллион",
  "октосептагинтиллион",
  "новемсептагинтиллион",
  "октогинтиллион",
  "уноктогинтиллион",
  "дуооктогинтиллион",
  "треоктогинтиллион",
  // 10^252
  "кватороктогинтиллион",
  "квиноктогинтиллион",
  "сексоктогинтиллион",
  "септоктогинтиллион",
  "октооктогинтиллион",
  "новемоктогинтиллион",
  "нонагинтиллион",
  "уннонагинтиллион",
  "дуононагинтиллион",
  "тренонагинтиллион",
  "кваторнонагинтиллион",
  "квиннонагинтиллион",
  "секснонагинтиллион",
  "септеннонагинтиллион",
  "октононагинтиллион",
  "новемнонагинтиллион",
  // 10^300
  "центиллион"
  // 10^303
], NI = {
  [I.NOMINATIVE]: ["", "ы"],
  [I.GENITIVE]: ["а", "ов"],
  [I.DATIVE]: ["у", "ам"],
  [I.ACCUSATIVE]: ["", "а"],
  [I.INSTRUMENTAL]: ["ом", "ами"],
  [I.PREPOSITIONAL]: ["е", "ах"]
}, AI = {
  [I.NOMINATIVE]: ["тысяча", "тысячи"],
  [I.GENITIVE]: ["тысячи", "тысяч"],
  [I.DATIVE]: ["тысяче", "тысячам"],
  [I.ACCUSATIVE]: ["тысячу", "тысячи"],
  [I.INSTRUMENTAL]: ["тысячей", "тысячами"],
  [I.PREPOSITIONAL]: ["тысяче", "тысячах"]
};
function eI(e, A, N) {
  let t = N, E = A === 0 ? 0 : 1;
  if ((N === I.NOMINATIVE || N === I.ACCUSATIVE) && A >= 1 && (t = I.GENITIVE, E = A === 1 ? 0 : 1), e === 0)
    return "";
  if (e === 1)
    return AI[t][E];
  const V = NI[t][E];
  return j[e - 2] + V;
}
const T = {
  MALE: "MALE",
  NEUTER: "NEUTER",
  FEMALE: "FEMALE"
}, tI = "минус", D = {
  numbers: {
    [I.NOMINATIVE]: [
      "ноль",
      {
        [T.MALE]: "один",
        [T.NEUTER]: "одно",
        [T.FEMALE]: "одна"
      },
      {
        [T.MALE]: "два",
        [T.NEUTER]: "два",
        [T.FEMALE]: "две"
      },
      "три",
      "четыре",
      "пять",
      "шесть",
      "семь",
      "восемь",
      "девять"
    ],
    [I.GENITIVE]: [
      "ноля",
      {
        [T.MALE]: "одного",
        [T.NEUTER]: "одного",
        [T.FEMALE]: "одной"
      },
      "двух",
      "трёх",
      "четырёх",
      "пяти",
      "шести",
      "семи",
      "восьми",
      "девяти"
    ],
    [I.DATIVE]: [
      "нолю",
      {
        [T.MALE]: "одному",
        [T.NEUTER]: "одному",
        [T.FEMALE]: "одной"
      },
      "двум",
      "трём",
      "четырём",
      "пяти",
      "шести",
      "семи",
      "восьми",
      "девяти"
    ],
    [I.ACCUSATIVE]: [
      "ноль",
      {
        [T.MALE]: "один",
        [T.NEUTER]: "одно",
        [T.FEMALE]: "одну"
      },
      {
        [T.MALE]: "два",
        [T.NEUTER]: "два",
        [T.FEMALE]: "две"
      },
      "три",
      "четыре",
      "пять",
      "шесть",
      "семь",
      "восемь",
      "девять"
    ],
    [I.INSTRUMENTAL]: [
      "нолём",
      {
        [T.MALE]: "одним",
        [T.NEUTER]: "одним",
        [T.FEMALE]: "одной"
      },
      "двумя",
      "тремя",
      "четырьмя",
      "пятью",
      "шестью",
      "семью",
      "восемью",
      "девятью"
    ],
    [I.PREPOSITIONAL]: [
      "ноле",
      {
        [T.MALE]: "одном",
        [T.NEUTER]: "одном",
        [T.FEMALE]: "одной"
      },
      "двух",
      "трёх",
      "четырёх",
      "пяти",
      "шести",
      "семи",
      "восьми",
      "девяти"
    ]
  },
  tenToNineteen: {
    [I.NOMINATIVE]: [
      "десять",
      "одиннадцать",
      "двенадцать",
      "тринадцать",
      "четырнадцать",
      "пятнадцать",
      "шестнадцать",
      "семнадцать",
      "восемнадцать",
      "девятнадцать"
    ],
    [I.GENITIVE]: [
      "десяти",
      "одиннадцати",
      "двенадцати",
      "тринадцати",
      "четырнадцати",
      "пятнадцати",
      "шестнадцати",
      "семнадцати",
      "восемнадцати",
      "девятнадцати"
    ],
    [I.DATIVE]: [
      "десяти",
      "одиннадцати",
      "двенадцати",
      "тринадцати",
      "четырнадцати",
      "пятнадцати",
      "шестнадцати",
      "семнадцати",
      "восемнадцати",
      "девятнадцати"
    ],
    [I.ACCUSATIVE]: [
      "десять",
      "одиннадцать",
      "двенадцать",
      "тринадцать",
      "четырнадцать",
      "пятнадцать",
      "шестнадцать",
      "семнадцать",
      "восемнадцать",
      "девятнадцать"
    ],
    [I.INSTRUMENTAL]: [
      "десятью",
      "одиннадцатью",
      "двенадцатью",
      "тринадцатью",
      "четырнадцатью",
      "пятнадцатью",
      "шестнадцатью",
      "семнадцатью",
      "восемнадцатью",
      "девятнадцатью"
    ],
    [I.PREPOSITIONAL]: [
      "десяти",
      "одиннадцати",
      "двенадцати",
      "тринадцати",
      "четырнадцати",
      "пятнадцати",
      "шестнадцати",
      "семнадцати",
      "восемнадцати",
      "девятнадцати"
    ]
  },
  tens: {
    [I.NOMINATIVE]: [
      "",
      "",
      "двадцать",
      "тридцать",
      "сорок",
      "пятьдесят",
      "шестьдесят",
      "семьдесят",
      "восемьдесят",
      "девяносто"
    ],
    [I.GENITIVE]: [
      "",
      "",
      "двадцати",
      "тридцати",
      "сорока",
      "пятидесяти",
      "шестидесяти",
      "семидесяти",
      "восьмидесяти",
      "девяноста"
    ],
    [I.DATIVE]: [
      "",
      "",
      "двадцати",
      "тридцати",
      "сорока",
      "пятидесяти",
      "шестидесяти",
      "семидесяти",
      "восьмидесяти",
      "девяноста"
    ],
    [I.ACCUSATIVE]: [
      "",
      "",
      "двадцать",
      "тридцать",
      "сорок",
      "пятьдесят",
      "шестьдесят",
      "семьдесят",
      "восемьдесят",
      "девяносто"
    ],
    [I.INSTRUMENTAL]: [
      "",
      "",
      "двадцатью",
      "тридцатью",
      "сорока",
      "пятьюдесятью",
      "шестьюдесятью",
      "семьюдесятью",
      "восемьюдесятью",
      "девяноста"
    ],
    [I.PREPOSITIONAL]: [
      "",
      "",
      "двадцати",
      "тридцати",
      "сорока",
      "пятидесяти",
      "шестидесяти",
      "семидесяти",
      "восьмидесяти",
      "девяноста"
    ]
  },
  hundreds: {
    [I.NOMINATIVE]: [
      "",
      "сто",
      "двести",
      "триста",
      "четыреста",
      "пятьсот",
      "шестьсот",
      "семьсот",
      "восемьсот",
      "девятьсот"
    ],
    [I.GENITIVE]: [
      "",
      "ста",
      "двухсот",
      "трёхсот",
      "четырёхсот",
      "пятисот",
      "шестисот",
      "семисот",
      "восьмисот",
      "девятисот"
    ],
    [I.DATIVE]: [
      "",
      "ста",
      "двумстам",
      "трёмстам",
      "четырёмстам",
      "пятистам",
      "шестистам",
      "семистам",
      "восьмистам",
      "девятистам"
    ],
    [I.ACCUSATIVE]: [
      "",
      "сто",
      "двести",
      "триста",
      "четыреста",
      "пятьсот",
      "шестьсот",
      "семьсот",
      "восемьсот",
      "девятьсот"
    ],
    [I.INSTRUMENTAL]: [
      "",
      "ста",
      "двумястами",
      "тремястами",
      "четырьмястами",
      "пятьюстами",
      "шестьюстами",
      "семьюстами",
      "восемьюстами",
      "девятьюстами"
    ],
    [I.PREPOSITIONAL]: [
      "",
      "ста",
      "двухстах",
      "трёхстах",
      "четырёхстах",
      "пятистах",
      "шестистах",
      "семистах",
      "восьмистах",
      "девятистах"
    ]
  }
};
function P(e, A = 0, N = I.NOMINATIVE) {
  let t = "", E = 2, V = !1;
  return e.some((r, n) => {
    E = 2, V = !1;
    const s = e.length - n, S = parseInt(r[0]), c = parseInt(r[1]), o = parseInt(r[2]);
    let O = "", M = "", i = "";
    if (S === 0 && c === 0 && o === 0)
      return V = !0, e.length === 1 ? (t = f(
        o,
        D.numbers,
        N,
        T.MALE
      ), E = 2, !0) : void 0;
    let L = T.MALE;
    s === 2 ? L = T.FEMALE : s === 1 && (A === 1 ? L = T.FEMALE : A === 2 && (L = T.NEUTER)), O = f(
      S,
      D.hundreds,
      N,
      L
    ), c === 1 ? M = f(
      o,
      D.tenToNineteen,
      N,
      L
    ) : (M = f(c, D.tens, N, L), i = f(
      o,
      D.numbers,
      N,
      L
    ), o >= 1 && o <= 4 && (o === 1 && (E = 0), o >= 2 && o <= 4 && (E = 1)));
    const m = eI(
      s - 1,
      E,
      N
    );
    o === 0 && (S > 0 || c > 0) && (i = "");
    const d = `${O} ${M} ${i} ${m}`.replace(/\s+/g, " ").trim();
    t += ` ${d}`;
  }), {
    result: t.trim(),
    unitNameForm: E,
    lastScaleIsZero: V
  };
}
const rI = (e, A) => e.length === A.length && e.every((N, t) => N === A[t]);
function nI(e) {
  const A = e.findIndex((t) => t !== "000");
  return e.slice(A);
}
function b(e, A, N = !1) {
  var E, V;
  let t = (E = e[A]) == null ? void 0 : E[N ? 1 : 0];
  return N === !0 && (A === I.NOMINATIVE || A === I.ACCUSATIVE) && (t = (V = e[I.GENITIVE]) == null ? void 0 : V[1]), t || "";
}
const z = [
  {
    [I.NOMINATIVE]: ["десятая", "десятые"],
    [I.GENITIVE]: ["десятой", "десятых"],
    [I.DATIVE]: ["десятой", "десятым"],
    [I.ACCUSATIVE]: ["десятую", "десятых"],
    [I.INSTRUMENTAL]: ["десятой", "десятыми"],
    [I.PREPOSITIONAL]: ["десятой", "десятых"]
  },
  {
    [I.NOMINATIVE]: ["сотая", "сотые"],
    [I.GENITIVE]: ["сотой", "сотых"],
    [I.DATIVE]: ["сотой", "сотым"],
    [I.ACCUSATIVE]: ["сотую", "сотых"],
    [I.INSTRUMENTAL]: ["сотой", "сотыми"],
    [I.PREPOSITIONAL]: ["сотой", "сотых"]
  }
], h = [
  "тысяч"
  // 'миллион',
  // 'миллиард',
  // ...
], VI = ["", "десяти", "сто"], w = {
  [I.NOMINATIVE]: ["ная", "ные"],
  [I.GENITIVE]: ["ной", "ных"],
  [I.DATIVE]: ["ной", "ным"],
  [I.ACCUSATIVE]: ["ную", "ные"],
  [I.INSTRUMENTAL]: ["ной", "ными"],
  [I.PREPOSITIONAL]: ["ной", "ных"]
}, sI = [
  {
    // 0
    [T.MALE]: {
      [I.NOMINATIVE]: ["нулевой", "нулевые"],
      [I.GENITIVE]: ["нулевого", "нулевых"],
      [I.DATIVE]: ["нулевому", "нулевым"],
      [I.ACCUSATIVE]: ["нулевой", "нулевые"],
      [I.INSTRUMENTAL]: ["нулевым", "нулевыми"],
      [I.PREPOSITIONAL]: ["нулевом", "нулевых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["нулевое", "нулевые"],
      [I.GENITIVE]: ["нулевого", "нулевых"],
      [I.DATIVE]: ["нулевому", "нулевым"],
      [I.ACCUSATIVE]: ["нулевое", "нулевые"],
      [I.INSTRUMENTAL]: ["нулевым", "нулевыми"],
      [I.PREPOSITIONAL]: ["нулевом", "нулевых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["нулевая", "нулевые"],
      [I.GENITIVE]: ["нулевой", "нулевых"],
      [I.DATIVE]: ["нулевой", "нулевым"],
      [I.ACCUSATIVE]: ["нулевую", "нулевые"],
      [I.INSTRUMENTAL]: ["нулевой", "нулевыми"],
      [I.PREPOSITIONAL]: ["нулевой", "нулевых"]
    }
  },
  {
    // 1
    [T.MALE]: {
      [I.NOMINATIVE]: ["первый", "первые"],
      [I.GENITIVE]: ["первого", "первых"],
      [I.DATIVE]: ["первому", "первым"],
      [I.ACCUSATIVE]: ["первый", "первые"],
      [I.INSTRUMENTAL]: ["первым", "первыми"],
      [I.PREPOSITIONAL]: ["первом", "первых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["первое", "первые"],
      [I.GENITIVE]: ["первого", "первых"],
      [I.DATIVE]: ["первому", "первым"],
      [I.ACCUSATIVE]: ["первое", "первые"],
      [I.INSTRUMENTAL]: ["первым", "первыми"],
      [I.PREPOSITIONAL]: ["первом", "первых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["первая", "первые"],
      [I.GENITIVE]: ["первой", "первых"],
      [I.DATIVE]: ["первой", "первым"],
      [I.ACCUSATIVE]: ["первую", "первые"],
      [I.INSTRUMENTAL]: ["первой", "первыми"],
      [I.PREPOSITIONAL]: ["первой", "первых"]
    }
  },
  {
    // 2
    [T.MALE]: {
      [I.NOMINATIVE]: ["второй", "вторые"],
      [I.GENITIVE]: ["второго", "вторых"],
      [I.DATIVE]: ["второму", "вторым"],
      [I.ACCUSATIVE]: ["второй", "вторые"],
      [I.INSTRUMENTAL]: ["вторым", "вторыми"],
      [I.PREPOSITIONAL]: ["втором", "вторых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["второе", "вторые"],
      [I.GENITIVE]: ["второго", "вторых"],
      [I.DATIVE]: ["второму", "вторым"],
      [I.ACCUSATIVE]: ["второе", "вторые"],
      [I.INSTRUMENTAL]: ["вторым", "вторыми"],
      [I.PREPOSITIONAL]: ["втором", "вторых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["вторая", "вторые"],
      [I.GENITIVE]: ["второй", "вторых"],
      [I.DATIVE]: ["второй", "вторым"],
      [I.ACCUSATIVE]: ["вторую", "вторые"],
      [I.INSTRUMENTAL]: ["второй", "вторыми"],
      [I.PREPOSITIONAL]: ["второй", "вторых"]
    }
  },
  {
    // 3
    [T.MALE]: {
      [I.NOMINATIVE]: ["третий", "третьи"],
      [I.GENITIVE]: ["третьего", "третьих"],
      [I.DATIVE]: ["третьему", "третьим"],
      [I.ACCUSATIVE]: ["третий", "третьи"],
      [I.INSTRUMENTAL]: ["третьим", "третьими"],
      [I.PREPOSITIONAL]: ["третьем", "третьих"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["третье", "третьи"],
      [I.GENITIVE]: ["третьего", "третьих"],
      [I.DATIVE]: ["третьему", "третьим"],
      [I.ACCUSATIVE]: ["третье", "третьи"],
      [I.INSTRUMENTAL]: ["третьим", "третьими"],
      [I.PREPOSITIONAL]: ["третьем", "третьих"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["третья", "третьи"],
      [I.GENITIVE]: ["третьей", "третьих"],
      [I.DATIVE]: ["третьей", "третьим"],
      [I.ACCUSATIVE]: ["третью", "третьи"],
      [I.INSTRUMENTAL]: ["третьей", "третьими"],
      [I.PREPOSITIONAL]: ["третьей", "третьих"]
    }
  },
  {
    // 4
    [T.MALE]: {
      [I.NOMINATIVE]: ["четвёртый", "четвёртые"],
      [I.GENITIVE]: ["четвёртого", "четвёртых"],
      [I.DATIVE]: ["четвёртому", "четвёртым"],
      [I.ACCUSATIVE]: ["четвёртый", "четвёртые"],
      [I.INSTRUMENTAL]: ["четвёртым", "четвёртыми"],
      [I.PREPOSITIONAL]: ["четвёртом", "четвёртых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["четвёртое", "четвёртые"],
      [I.GENITIVE]: ["четвёртого", "четвёртых"],
      [I.DATIVE]: ["четвёртому", "четвёртым"],
      [I.ACCUSATIVE]: ["четвёртое", "четвёртые"],
      [I.INSTRUMENTAL]: ["четвёртым", "четвёртыми"],
      [I.PREPOSITIONAL]: ["четвёртом", "четвёртых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["четвёртая", "четвёртые"],
      [I.GENITIVE]: ["четвёртой", "четвёртых"],
      [I.DATIVE]: ["четвёртой", "четвёртым"],
      [I.ACCUSATIVE]: ["четвёртую", "четвёртые"],
      [I.INSTRUMENTAL]: ["четвёртой", "четвёртыми"],
      [I.PREPOSITIONAL]: ["четвёртой", "четвёртых"]
    }
  },
  {
    // 5
    [T.MALE]: {
      [I.NOMINATIVE]: ["пятый", "пятые"],
      [I.GENITIVE]: ["пятого", "пятых"],
      [I.DATIVE]: ["пятому", "пятым"],
      [I.ACCUSATIVE]: ["пятый", "пятые"],
      [I.INSTRUMENTAL]: ["пятым", "пятыми"],
      [I.PREPOSITIONAL]: ["пятом", "пятых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["пятое", "пятые"],
      [I.GENITIVE]: ["пятого", "пятых"],
      [I.DATIVE]: ["пятому", "пятым"],
      [I.ACCUSATIVE]: ["пятое", "пятые"],
      [I.INSTRUMENTAL]: ["пятым", "пятыми"],
      [I.PREPOSITIONAL]: ["пятом", "пятых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["пятая", "пятые"],
      [I.GENITIVE]: ["пятой", "пятых"],
      [I.DATIVE]: ["пятой", "пятым"],
      [I.ACCUSATIVE]: ["пятую", "пятые"],
      [I.INSTRUMENTAL]: ["пятой", "пятыми"],
      [I.PREPOSITIONAL]: ["пятой", "пятых"]
    }
  },
  {
    // 6
    [T.MALE]: {
      [I.NOMINATIVE]: ["шестой", "шестые"],
      [I.GENITIVE]: ["шестого", "шестых"],
      [I.DATIVE]: ["шестому", "шестым"],
      [I.ACCUSATIVE]: ["шестой", "шестые"],
      [I.INSTRUMENTAL]: ["шестым", "шестыми"],
      [I.PREPOSITIONAL]: ["шестом", "шестых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["шестое", "шестые"],
      [I.GENITIVE]: ["шестого", "шестых"],
      [I.DATIVE]: ["шестому", "шестым"],
      [I.ACCUSATIVE]: ["шестое", "шестые"],
      [I.INSTRUMENTAL]: ["шестым", "шестыми"],
      [I.PREPOSITIONAL]: ["шестом", "шестых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["шестая", "шестые"],
      [I.GENITIVE]: ["шестой", "шестых"],
      [I.DATIVE]: ["шестой", "шестым"],
      [I.ACCUSATIVE]: ["шестую", "шестые"],
      [I.INSTRUMENTAL]: ["шестой", "шестыми"],
      [I.PREPOSITIONAL]: ["шестой", "шестых"]
    }
  },
  {
    // 7
    [T.MALE]: {
      [I.NOMINATIVE]: ["седьмой", "седьмые"],
      [I.GENITIVE]: ["седьмого", "седьмых"],
      [I.DATIVE]: ["седьмому", "седьмым"],
      [I.ACCUSATIVE]: ["седьмой", "седьмые"],
      [I.INSTRUMENTAL]: ["седьмым", "седьмыми"],
      [I.PREPOSITIONAL]: ["седьмом", "седьмых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["седьмое", "седьмые"],
      [I.GENITIVE]: ["седьмого", "седьмых"],
      [I.DATIVE]: ["седьмому", "седьмым"],
      [I.ACCUSATIVE]: ["седьмое", "седьмые"],
      [I.INSTRUMENTAL]: ["седьмым", "седьмыми"],
      [I.PREPOSITIONAL]: ["седьмом", "седьмых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["седьмая", "седьмые"],
      [I.GENITIVE]: ["седьмой", "седьмых"],
      [I.DATIVE]: ["седьмой", "седьмым"],
      [I.ACCUSATIVE]: ["седьмую", "седьмые"],
      [I.INSTRUMENTAL]: ["седьмой", "седьмыми"],
      [I.PREPOSITIONAL]: ["седьмой", "седьмых"]
    }
  },
  {
    // 8
    [T.MALE]: {
      [I.NOMINATIVE]: ["восьмой", "восьмые"],
      [I.GENITIVE]: ["восьмого", "восьмых"],
      [I.DATIVE]: ["восьмому", "восьмым"],
      [I.ACCUSATIVE]: ["восьмой", "восьмые"],
      [I.INSTRUMENTAL]: ["восьмым", "восьмыми"],
      [I.PREPOSITIONAL]: ["восьмом", "восьмых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["восьмое", "восьмые"],
      [I.GENITIVE]: ["восьмого", "восьмых"],
      [I.DATIVE]: ["восьмому", "восьмым"],
      [I.ACCUSATIVE]: ["восьмое", "восьмые"],
      [I.INSTRUMENTAL]: ["восьмым", "восьмыми"],
      [I.PREPOSITIONAL]: ["восьмом", "восьмых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["восьмая", "восьмые"],
      [I.GENITIVE]: ["восьмой", "восьмых"],
      [I.DATIVE]: ["восьмой", "восьмым"],
      [I.ACCUSATIVE]: ["восьмую", "восьмые"],
      [I.INSTRUMENTAL]: ["восьмой", "восьмыми"],
      [I.PREPOSITIONAL]: ["восьмой", "восьмых"]
    }
  },
  {
    // 9
    [T.MALE]: {
      [I.NOMINATIVE]: ["девятый", "девятые"],
      [I.GENITIVE]: ["девятого", "девятых"],
      [I.DATIVE]: ["девятому", "девятым"],
      [I.ACCUSATIVE]: ["девятый", "девятые"],
      [I.INSTRUMENTAL]: ["девятым", "девятыми"],
      [I.PREPOSITIONAL]: ["девятом", "девятых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["девятое", "девятые"],
      [I.GENITIVE]: ["девятого", "девятых"],
      [I.DATIVE]: ["девятому", "девятым"],
      [I.ACCUSATIVE]: ["девятое", "девятые"],
      [I.INSTRUMENTAL]: ["девятым", "девятыми"],
      [I.PREPOSITIONAL]: ["девятом", "девятых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["девятая", "девятые"],
      [I.GENITIVE]: ["девятой", "девятых"],
      [I.DATIVE]: ["девятой", "девятым"],
      [I.ACCUSATIVE]: ["девятую", "девятые"],
      [I.INSTRUMENTAL]: ["девятой", "девятыми"],
      [I.PREPOSITIONAL]: ["девятой", "девятых"]
    }
  }
], cI = [
  {
    // 0 (пусто)
  },
  {
    // 100
    [T.MALE]: {
      [I.NOMINATIVE]: ["сотый", "сотые"],
      [I.GENITIVE]: ["сотого", "сотых"],
      [I.DATIVE]: ["сотому", "сотым"],
      [I.ACCUSATIVE]: ["сотый", "сотые"],
      [I.INSTRUMENTAL]: ["сотым", "сотыми"],
      [I.PREPOSITIONAL]: ["сотом", "сотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["сотое", "сотые"],
      [I.GENITIVE]: ["сотого", "сотых"],
      [I.DATIVE]: ["сотому", "сотым"],
      [I.ACCUSATIVE]: ["сотое", "сотые"],
      [I.INSTRUMENTAL]: ["сотым", "сотыми"],
      [I.PREPOSITIONAL]: ["сотом", "сотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["сотая", "сотые"],
      [I.GENITIVE]: ["сотой", "сотых"],
      [I.DATIVE]: ["сотой", "сотым"],
      [I.ACCUSATIVE]: ["сотую", "сотые"],
      [I.INSTRUMENTAL]: ["сотой", "сотыми"],
      [I.PREPOSITIONAL]: ["сотой", "сотых"]
    }
  },
  {
    // 200
    [T.MALE]: {
      [I.NOMINATIVE]: ["двухсотый", "двухсотые"],
      [I.GENITIVE]: ["двухсотого", "двухсотых"],
      [I.DATIVE]: ["двухсотому", "двухсотым"],
      [I.ACCUSATIVE]: ["двухсотый", "двухсотые"],
      [I.INSTRUMENTAL]: ["двухсотым", "двухсотыми"],
      [I.PREPOSITIONAL]: ["двухсотом", "двухсотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["двухсотое", "двухсотые"],
      [I.GENITIVE]: ["двухсотого", "двухсотых"],
      [I.DATIVE]: ["двухсотому", "двухсотым"],
      [I.ACCUSATIVE]: ["двухсотое", "двухсотые"],
      [I.INSTRUMENTAL]: ["двухсотым", "двухсотыми"],
      [I.PREPOSITIONAL]: ["двухсотом", "двухсотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["двухсотая", "двухсотые"],
      [I.GENITIVE]: ["двухсотой", "двухсотых"],
      [I.DATIVE]: ["двухсотой", "двухсотым"],
      [I.ACCUSATIVE]: ["двухсотую", "двухсотые"],
      [I.INSTRUMENTAL]: ["двухсотой", "двухсотыми"],
      [I.PREPOSITIONAL]: ["двухсотой", "двухсотых"]
    }
  },
  {
    // 300
    [T.MALE]: {
      [I.NOMINATIVE]: ["трёхсотый", "трёхсотые"],
      [I.GENITIVE]: ["трёхсотого", "трёхсотых"],
      [I.DATIVE]: ["трёхсотому", "трёхсотым"],
      [I.ACCUSATIVE]: ["трёхсотый", "трёхсотые"],
      [I.INSTRUMENTAL]: ["трёхсотым", "трёхсотыми"],
      [I.PREPOSITIONAL]: ["трёхсотом", "трёхсотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["трёхсотое", "трёхсотые"],
      [I.GENITIVE]: ["трёхсотого", "трёхсотых"],
      [I.DATIVE]: ["трёхсотому", "трёхсотым"],
      [I.ACCUSATIVE]: ["трёхсотое", "трёхсотые"],
      [I.INSTRUMENTAL]: ["трёхсотым", "трёхсотыми"],
      [I.PREPOSITIONAL]: ["трёхсотом", "трёхсотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["трёхсотая	", "трёхсотые"],
      [I.GENITIVE]: ["трёхсотой", "трёхсотых"],
      [I.DATIVE]: ["трёхсотой", "трёхсотым"],
      [I.ACCUSATIVE]: ["трёхсотую", "трёхсотые"],
      [I.INSTRUMENTAL]: ["трёхсотой", "трёхсотыми"],
      [I.PREPOSITIONAL]: ["трёхсотой", "трёхсотых"]
    }
  },
  {
    // 400
    [T.MALE]: {
      [I.NOMINATIVE]: ["четырёхсотый", "четырёхсотые"],
      [I.GENITIVE]: ["четырёхсотого", "четырёхсотых"],
      [I.DATIVE]: ["четырёхсотому", "четырёхсотым"],
      [I.ACCUSATIVE]: ["четырёхсотый", "четырёхсотые"],
      [I.INSTRUMENTAL]: ["четырёхсотым", "четырёхсотыми"],
      [I.PREPOSITIONAL]: ["четырёхсотом", "четырёхсотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["четырёхсотое", "четырёхсотые"],
      [I.GENITIVE]: ["четырёхсотого", "четырёхсотых"],
      [I.DATIVE]: ["четырёхсотому", "четырёхсотым"],
      [I.ACCUSATIVE]: ["четырёхсотое", "четырёхсотые"],
      [I.INSTRUMENTAL]: ["четырёхсотым", "четырёхсотыми"],
      [I.PREPOSITIONAL]: ["четырёхсотом", "четырёхсотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["четырёхсотая", "четырёхсотые"],
      [I.GENITIVE]: ["четырёхсотой", "четырёхсотых"],
      [I.DATIVE]: ["четырёхсотой", "четырёхсотым"],
      [I.ACCUSATIVE]: ["четырёхсотую", "четырёхсотые"],
      [I.INSTRUMENTAL]: ["четырёхсотой", "четырёхсотыми"],
      [I.PREPOSITIONAL]: ["четырёхсотой", "четырёхсотых"]
    }
  },
  {
    // 500
    [T.MALE]: {
      [I.NOMINATIVE]: ["пятисотый", "пятисотые"],
      [I.GENITIVE]: ["пятисотого", "пятисотых"],
      [I.DATIVE]: ["пятисотому", "пятисотым"],
      [I.ACCUSATIVE]: ["пятисотый", "пятисотые"],
      [I.INSTRUMENTAL]: ["пятисотым", "пятисотыми"],
      [I.PREPOSITIONAL]: ["пятисотом", "пятисотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["пятисотое", "пятисотые"],
      [I.GENITIVE]: ["пятисотого", "пятисотых"],
      [I.DATIVE]: ["пятисотому", "пятисотым"],
      [I.ACCUSATIVE]: ["пятисотое", "пятисотые"],
      [I.INSTRUMENTAL]: ["пятисотым", "пятисотыми"],
      [I.PREPOSITIONAL]: ["пятисотом", "пятисотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["пятисотая", "пятисотые"],
      [I.GENITIVE]: ["пятисотой", "пятисотых"],
      [I.DATIVE]: ["пятисотой", "пятисотым"],
      [I.ACCUSATIVE]: ["пятисотую", "пятисотые"],
      [I.INSTRUMENTAL]: ["пятисотой", "пятисотыми"],
      [I.PREPOSITIONAL]: ["пятисотой", "пятисотых"]
    }
  },
  {
    // 600
    [T.MALE]: {
      [I.NOMINATIVE]: ["шестисотый", "шестисотые"],
      [I.GENITIVE]: ["шестисотого", "шестисотых"],
      [I.DATIVE]: ["шестисотому", "шестисотым"],
      [I.ACCUSATIVE]: ["шестисотый", "шестисотые"],
      [I.INSTRUMENTAL]: ["шестисотым", "шестисотыми"],
      [I.PREPOSITIONAL]: ["шестисотом", "шестисотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["шестисотое", "шестисотые"],
      [I.GENITIVE]: ["шестисотого", "шестисотых"],
      [I.DATIVE]: ["шестисотому", "шестисотым"],
      [I.ACCUSATIVE]: ["шестисотое", "шестисотые"],
      [I.INSTRUMENTAL]: ["шестисотым", "шестисотыми"],
      [I.PREPOSITIONAL]: ["шестисотом", "шестисотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["шестисотая", "шестисотые"],
      [I.GENITIVE]: ["шестисотой", "шестисотых"],
      [I.DATIVE]: ["шестисотой", "шестисотым"],
      [I.ACCUSATIVE]: ["шестисотую", "шестисотые"],
      [I.INSTRUMENTAL]: ["шестисотой", "шестисотыми"],
      [I.PREPOSITIONAL]: ["шестисотой", "шестисотых"]
    }
  },
  {
    // 700
    [T.MALE]: {
      [I.NOMINATIVE]: ["семисотый", "семисотые"],
      [I.GENITIVE]: ["семисотого", "семисотых"],
      [I.DATIVE]: ["семисотому", "семисотым"],
      [I.ACCUSATIVE]: ["семисотый", "семисотые"],
      [I.INSTRUMENTAL]: ["семисотым", "семисотыми"],
      [I.PREPOSITIONAL]: ["семисотом", "семисотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["семисотое", "семисотые"],
      [I.GENITIVE]: ["семисотого", "семисотых"],
      [I.DATIVE]: ["семисотому", "семисотым"],
      [I.ACCUSATIVE]: ["семисотое", "семисотые"],
      [I.INSTRUMENTAL]: ["семисотым", "семисотыми"],
      [I.PREPOSITIONAL]: ["семисотом", "семисотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["семисотая", "семисотые"],
      [I.GENITIVE]: ["семисотой", "семисотых"],
      [I.DATIVE]: ["семисотой", "семисотым"],
      [I.ACCUSATIVE]: ["семисотую", "семисотые"],
      [I.INSTRUMENTAL]: ["семисотой", "семисотыми"],
      [I.PREPOSITIONAL]: ["семисотой", "семисотых"]
    }
  },
  {
    // 800
    [T.MALE]: {
      [I.NOMINATIVE]: ["восьмисотый", "восьмисотые"],
      [I.GENITIVE]: ["восьмисотого", "восьмисотых"],
      [I.DATIVE]: ["восьмисотому", "восьмисотым"],
      [I.ACCUSATIVE]: ["восьмисотый", "восьмисотые"],
      [I.INSTRUMENTAL]: ["восьмисотым", "восьмисотыми"],
      [I.PREPOSITIONAL]: ["восьмисотом", "восьмисотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["восьмисотое", "восьмисотые"],
      [I.GENITIVE]: ["восьмисотого", "восьмисотых"],
      [I.DATIVE]: ["восьмисотому", "восьмисотым"],
      [I.ACCUSATIVE]: ["восьмисотое", "восьмисотые"],
      [I.INSTRUMENTAL]: ["восьмисотым", "восьмисотыми"],
      [I.PREPOSITIONAL]: ["восьмисотом", "восьмисотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["восьмисотая", "восьмисотые"],
      [I.GENITIVE]: ["восьмисотой", "восьмисотых"],
      [I.DATIVE]: ["восьмисотой", "восьмисотым"],
      [I.ACCUSATIVE]: ["восьмисотую", "восьмисотые"],
      [I.INSTRUMENTAL]: ["восьмисотой", "восьмисотыми"],
      [I.PREPOSITIONAL]: ["восьмисотой", "восьмисотых"]
    }
  },
  {
    // 900
    [T.MALE]: {
      [I.NOMINATIVE]: ["девятисотый", "девятисотые"],
      [I.GENITIVE]: ["девятисотого", "девятисотых"],
      [I.DATIVE]: ["девятисотому", "девятисотым"],
      [I.ACCUSATIVE]: ["девятисотый", "девятисотые"],
      [I.INSTRUMENTAL]: ["девятисотым", "девятисотыми"],
      [I.PREPOSITIONAL]: ["девятисотом", "девятисотых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["девятисотое", "девятисотые"],
      [I.GENITIVE]: ["девятисотого", "девятисотых"],
      [I.DATIVE]: ["девятисотому", "девятисотым"],
      [I.ACCUSATIVE]: ["девятисотое", "девятисотые"],
      [I.INSTRUMENTAL]: ["девятисотым", "девятисотыми"],
      [I.PREPOSITIONAL]: ["девятисотом", "девятисотых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["девятисотая", "девятисотые"],
      [I.GENITIVE]: ["девятисотой", "девятисотых"],
      [I.DATIVE]: ["девятисотой", "девятисотым"],
      [I.ACCUSATIVE]: ["девятисотую", "девятисотые"],
      [I.INSTRUMENTAL]: ["девятисотой", "девятисотыми"],
      [I.PREPOSITIONAL]: ["девятисотой", "девятисотых"]
    }
  }
], oI = [
  {
    // 10
    [T.MALE]: {
      [I.NOMINATIVE]: ["десятый", "десятые"],
      [I.GENITIVE]: ["десятого", "десятых"],
      [I.DATIVE]: ["десятому", "десятым"],
      [I.ACCUSATIVE]: ["десятый", "десятые"],
      [I.INSTRUMENTAL]: ["десятым", "десятыми"],
      [I.PREPOSITIONAL]: ["десятом", "десятых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["десятое", "десятые"],
      [I.GENITIVE]: ["десятого", "десятых"],
      [I.DATIVE]: ["десятому", "десятым"],
      [I.ACCUSATIVE]: ["десятое", "десятые"],
      [I.INSTRUMENTAL]: ["десятым", "десятыми"],
      [I.PREPOSITIONAL]: ["десятом", "десятых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["десятая", "десятые"],
      [I.GENITIVE]: ["десятой", "десятых"],
      [I.DATIVE]: ["десятой", "десятым"],
      [I.ACCUSATIVE]: ["десятую", "десятые"],
      [I.INSTRUMENTAL]: ["десятой", "десятыми"],
      [I.PREPOSITIONAL]: ["десятой", "десятых"]
    }
  },
  {
    // 11
    [T.MALE]: {
      [I.NOMINATIVE]: ["одиннадцатый", "одиннадцатые"],
      [I.GENITIVE]: ["одиннадцатого", "одиннадцатых"],
      [I.DATIVE]: ["одиннадцатому", "одиннадцатым"],
      [I.ACCUSATIVE]: ["одиннадцатый", "одиннадцатые"],
      [I.INSTRUMENTAL]: ["одиннадцатым", "одиннадцатыми"],
      [I.PREPOSITIONAL]: ["одиннадцатом", "одиннадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["одиннадцатое", "одиннадцатые"],
      [I.GENITIVE]: ["одиннадцатого", "одиннадцатых"],
      [I.DATIVE]: ["одиннадцатому", "одиннадцатым"],
      [I.ACCUSATIVE]: ["одиннадцатое", "одиннадцатые"],
      [I.INSTRUMENTAL]: ["одиннадцатым", "одиннадцатыми"],
      [I.PREPOSITIONAL]: ["одиннадцатом", "одиннадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["одиннадцатая", "одиннадцатые"],
      [I.GENITIVE]: ["одиннадцатой", "одиннадцатых"],
      [I.DATIVE]: ["одиннадцатой", "одиннадцатым"],
      [I.ACCUSATIVE]: ["одиннадцатую", "одиннадцатые"],
      [I.INSTRUMENTAL]: ["одиннадцатой", "одиннадцатыми"],
      [I.PREPOSITIONAL]: ["одиннадцатой", "одиннадцатых"]
    }
  },
  {
    // 12
    [T.MALE]: {
      [I.NOMINATIVE]: ["двенадцатый", "двенадцатые"],
      [I.GENITIVE]: ["двенадцатого", "двенадцатых"],
      [I.DATIVE]: ["двенадцатому", "двенадцатым"],
      [I.ACCUSATIVE]: ["двенадцатый", "двенадцатые"],
      [I.INSTRUMENTAL]: ["двенадцатым", "двенадцатыми"],
      [I.PREPOSITIONAL]: ["двенадцатом", "двенадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["двенадцатое", "двенадцатые"],
      [I.GENITIVE]: ["двенадцатого", "двенадцатых"],
      [I.DATIVE]: ["двенадцатому", "двенадцатым"],
      [I.ACCUSATIVE]: ["двенадцатое", "двенадцатые"],
      [I.INSTRUMENTAL]: ["двенадцатым", "двенадцатыми"],
      [I.PREPOSITIONAL]: ["двенадцатом", "двенадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["двенадцатая", "двенадцатые"],
      [I.GENITIVE]: ["двенадцатой", "двенадцатых"],
      [I.DATIVE]: ["двенадцатой", "двенадцатым"],
      [I.ACCUSATIVE]: ["двенадцатую", "двенадцатые"],
      [I.INSTRUMENTAL]: ["двенадцатой", "двенадцатыми"],
      [I.PREPOSITIONAL]: ["двенадцатой", "двенадцатых"]
    }
  },
  {
    // 13
    [T.MALE]: {
      [I.NOMINATIVE]: ["тринадцатый", "тринадцатые"],
      [I.GENITIVE]: ["тринадцатого", "тринадцатых"],
      [I.DATIVE]: ["тринадцатому", "тринадцатым"],
      [I.ACCUSATIVE]: ["тринадцатый", "тринадцатые"],
      [I.INSTRUMENTAL]: ["тринадцатым", "тринадцатыми"],
      [I.PREPOSITIONAL]: ["тринадцатом", "тринадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["тринадцатое", "тринадцатые"],
      [I.GENITIVE]: ["тринадцатого", "тринадцатых"],
      [I.DATIVE]: ["тринадцатому", "тринадцатым"],
      [I.ACCUSATIVE]: ["тринадцатое", "тринадцатые"],
      [I.INSTRUMENTAL]: ["тринадцатым", "тринадцатыми"],
      [I.PREPOSITIONAL]: ["тринадцатом", "тринадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["тринадцатая", "тринадцатые"],
      [I.GENITIVE]: ["тринадцатой", "тринадцатых"],
      [I.DATIVE]: ["тринадцатой", "тринадцатым"],
      [I.ACCUSATIVE]: ["тринадцатую", "тринадцатые"],
      [I.INSTRUMENTAL]: ["тринадцатой", "тринадцатыми"],
      [I.PREPOSITIONAL]: ["тринадцатой", "тринадцатых"]
    }
  },
  {
    // 14
    [T.MALE]: {
      [I.NOMINATIVE]: ["четырнадцатый", "четырнадцатые"],
      [I.GENITIVE]: ["четырнадцатого", "четырнадцатых"],
      [I.DATIVE]: ["четырнадцатому", "четырнадцатым"],
      [I.ACCUSATIVE]: ["четырнадцатый", "четырнадцатые"],
      [I.INSTRUMENTAL]: ["четырнадцатым", "четырнадцатыми"],
      [I.PREPOSITIONAL]: ["четырнадцатом", "четырнадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["четырнадцатое", "четырнадцатые"],
      [I.GENITIVE]: ["четырнадцатого", "четырнадцатых"],
      [I.DATIVE]: ["четырнадцатому", "четырнадцатым"],
      [I.ACCUSATIVE]: ["четырнадцатое", "четырнадцатые"],
      [I.INSTRUMENTAL]: ["четырнадцатым", "четырнадцатыми"],
      [I.PREPOSITIONAL]: ["четырнадцатом", "четырнадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["четырнадцатая", "четырнадцатые"],
      [I.GENITIVE]: ["четырнадцатой", "четырнадцатых"],
      [I.DATIVE]: ["четырнадцатой", "четырнадцатым"],
      [I.ACCUSATIVE]: ["четырнадцатую", "четырнадцатые"],
      [I.INSTRUMENTAL]: ["четырнадцатой", "четырнадцатыми"],
      [I.PREPOSITIONAL]: ["четырнадцатой", "четырнадцатых"]
    }
  },
  {
    // 15
    [T.MALE]: {
      [I.NOMINATIVE]: ["пятнадцатый", "пятнадцатые"],
      [I.GENITIVE]: ["пятнадцатого", "пятнадцатых"],
      [I.DATIVE]: ["пятнадцатому", "пятнадцатым"],
      [I.ACCUSATIVE]: ["пятнадцатый", "пятнадцатые"],
      [I.INSTRUMENTAL]: ["пятнадцатым", "пятнадцатыми"],
      [I.PREPOSITIONAL]: ["пятнадцатом", "пятнадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["пятнадцатое", "пятнадцатые"],
      [I.GENITIVE]: ["пятнадцатого", "пятнадцатых"],
      [I.DATIVE]: ["пятнадцатому", "пятнадцатым"],
      [I.ACCUSATIVE]: ["пятнадцатое", "пятнадцатые"],
      [I.INSTRUMENTAL]: ["пятнадцатым", "пятнадцатыми"],
      [I.PREPOSITIONAL]: ["пятнадцатом", "пятнадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["пятнадцатая", "пятнадцатые"],
      [I.GENITIVE]: ["пятнадцатой", "пятнадцатых"],
      [I.DATIVE]: ["пятнадцатой", "пятнадцатым"],
      [I.ACCUSATIVE]: ["пятнадцатую", "пятнадцатые"],
      [I.INSTRUMENTAL]: ["пятнадцатой", "пятнадцатыми"],
      [I.PREPOSITIONAL]: ["пятнадцатой", "пятнадцатых"]
    }
  },
  {
    // 16
    [T.MALE]: {
      [I.NOMINATIVE]: ["шестнадцатый", "шестнадцатые"],
      [I.GENITIVE]: ["шестнадцатого", "шестнадцатых"],
      [I.DATIVE]: ["шестнадцатому", "шестнадцатым"],
      [I.ACCUSATIVE]: ["шестнадцатый", "шестнадцатые"],
      [I.INSTRUMENTAL]: ["шестнадцатым", "шестнадцатыми"],
      [I.PREPOSITIONAL]: ["шестнадцатом", "шестнадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["шестнадцатое", "шестнадцатые"],
      [I.GENITIVE]: ["шестнадцатого", "шестнадцатых"],
      [I.DATIVE]: ["шестнадцатому", "шестнадцатым"],
      [I.ACCUSATIVE]: ["шестнадцатое", "шестнадцатые"],
      [I.INSTRUMENTAL]: ["шестнадцатым", "шестнадцатыми"],
      [I.PREPOSITIONAL]: ["шестнадцатом", "шестнадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["шестнадцатая", "шестнадцатые"],
      [I.GENITIVE]: ["шестнадцатой", "шестнадцатых"],
      [I.DATIVE]: ["шестнадцатой", "шестнадцатым"],
      [I.ACCUSATIVE]: ["шестнадцатую", "шестнадцатые"],
      [I.INSTRUMENTAL]: ["шестнадцатой", "шестнадцатыми"],
      [I.PREPOSITIONAL]: ["шестнадцатой", "шестнадцатых"]
    }
  },
  {
    // 17
    [T.MALE]: {
      [I.NOMINATIVE]: ["семнадцатый", "семнадцатые"],
      [I.GENITIVE]: ["семнадцатого", "семнадцатых"],
      [I.DATIVE]: ["семнадцатому", "семнадцатым"],
      [I.ACCUSATIVE]: ["семнадцатый", "семнадцатые"],
      [I.INSTRUMENTAL]: ["семнадцатым", "семнадцатыми"],
      [I.PREPOSITIONAL]: ["семнадцатом", "семнадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["семнадцатое", "семнадцатые"],
      [I.GENITIVE]: ["семнадцатого", "семнадцатых"],
      [I.DATIVE]: ["семнадцатому", "семнадцатым"],
      [I.ACCUSATIVE]: ["семнадцатое", "семнадцатые"],
      [I.INSTRUMENTAL]: ["семнадцатым", "семнадцатыми"],
      [I.PREPOSITIONAL]: ["семнадцатом", "семнадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["семнадцатая", "семнадцатые"],
      [I.GENITIVE]: ["семнадцатой", "семнадцатых"],
      [I.DATIVE]: ["семнадцатой", "семнадцатым"],
      [I.ACCUSATIVE]: ["семнадцатую", "семнадцатые"],
      [I.INSTRUMENTAL]: ["семнадцатой", "семнадцатыми"],
      [I.PREPOSITIONAL]: ["семнадцатой", "семнадцатых"]
    }
  },
  {
    // 18
    [T.MALE]: {
      [I.NOMINATIVE]: ["восемнадцатый", "восемнадцатые"],
      [I.GENITIVE]: ["восемнадцатого", "восемнадцатых"],
      [I.DATIVE]: ["восемнадцатому", "восемнадцатым"],
      [I.ACCUSATIVE]: ["восемнадцатый", "восемнадцатые"],
      [I.INSTRUMENTAL]: ["восемнадцатым", "восемнадцатыми"],
      [I.PREPOSITIONAL]: ["восемнадцатом", "восемнадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["восемнадцатое", "восемнадцатые"],
      [I.GENITIVE]: ["восемнадцатого", "восемнадцатых"],
      [I.DATIVE]: ["восемнадцатому", "восемнадцатым"],
      [I.ACCUSATIVE]: ["восемнадцатое", "восемнадцатые"],
      [I.INSTRUMENTAL]: ["восемнадцатым", "восемнадцатыми"],
      [I.PREPOSITIONAL]: ["восемнадцатом", "восемнадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["восемнадцатая", "восемнадцатые"],
      [I.GENITIVE]: ["восемнадцатой", "восемнадцатых"],
      [I.DATIVE]: ["восемнадцатой", "восемнадцатым"],
      [I.ACCUSATIVE]: ["восемнадцатую", "восемнадцатые"],
      [I.INSTRUMENTAL]: ["восемнадцатой", "восемнадцатыми"],
      [I.PREPOSITIONAL]: ["восемнадцатой", "восемнадцатых"]
    }
  },
  {
    // 19
    [T.MALE]: {
      [I.NOMINATIVE]: ["девятнадцатый", "девятнадцатые"],
      [I.GENITIVE]: ["девятнадцатого", "девятнадцатых"],
      [I.DATIVE]: ["девятнадцатому", "девятнадцатым"],
      [I.ACCUSATIVE]: ["девятнадцатый", "девятнадцатые"],
      [I.INSTRUMENTAL]: ["девятнадцатым", "девятнадцатыми"],
      [I.PREPOSITIONAL]: ["девятнадцатом", "девятнадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["девятнадцатое", "девятнадцатые"],
      [I.GENITIVE]: ["девятнадцатого", "девятнадцатых"],
      [I.DATIVE]: ["девятнадцатому", "девятнадцатым"],
      [I.ACCUSATIVE]: ["девятнадцатое", "девятнадцатые"],
      [I.INSTRUMENTAL]: ["девятнадцатым", "девятнадцатыми"],
      [I.PREPOSITIONAL]: ["девятнадцатом", "девятнадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["девятнадцатая", "девятнадцатые"],
      [I.GENITIVE]: ["девятнадцатой", "девятнадцатых"],
      [I.DATIVE]: ["девятнадцатой", "девятнадцатым"],
      [I.ACCUSATIVE]: ["девятнадцатую", "девятнадцатые"],
      [I.INSTRUMENTAL]: ["девятнадцатой", "девятнадцатыми"],
      [I.PREPOSITIONAL]: ["девятнадцатой", "девятнадцатых"]
    }
  }
], SI = [
  {
    // 0 (пусто)
  },
  {
    // 10 (пусто)
  },
  {
    // 20
    [T.MALE]: {
      [I.NOMINATIVE]: ["двадцатый", "двадцатые"],
      [I.GENITIVE]: ["двадцатого", "двадцатых"],
      [I.DATIVE]: ["двадцатому", "двадцатым"],
      [I.ACCUSATIVE]: ["двадцатый", "двадцатые"],
      [I.INSTRUMENTAL]: ["двадцатым", "двадцатыми"],
      [I.PREPOSITIONAL]: ["двадцатом", "двадцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["двадцатое", "двадцатые"],
      [I.GENITIVE]: ["двадцатого", "двадцатых"],
      [I.DATIVE]: ["двадцатому", "двадцатым"],
      [I.ACCUSATIVE]: ["двадцатое", "двадцатые"],
      [I.INSTRUMENTAL]: ["двадцатым", "двадцатыми"],
      [I.PREPOSITIONAL]: ["двадцатом", "двадцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["двадцатая", "двадцатые"],
      [I.GENITIVE]: ["двадцатой", "двадцатых"],
      [I.DATIVE]: ["двадцатой", "двадцатым"],
      [I.ACCUSATIVE]: ["двадцатую", "двадцатые"],
      [I.INSTRUMENTAL]: ["двадцатой", "двадцатыми"],
      [I.PREPOSITIONAL]: ["двадцатой", "двадцатых"]
    }
  },
  {
    // 30
    [T.MALE]: {
      [I.NOMINATIVE]: ["тридцатый", "тридцатые"],
      [I.GENITIVE]: ["тридцатого", "тридцатых"],
      [I.DATIVE]: ["тридцатому", "тридцатым"],
      [I.ACCUSATIVE]: ["тридцатый", "тридцатые"],
      [I.INSTRUMENTAL]: ["тридцатым", "тридцатыми"],
      [I.PREPOSITIONAL]: ["тридцатом", "тридцатых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["тридцатое", "тридцатые"],
      [I.GENITIVE]: ["тридцатого", "тридцатых"],
      [I.DATIVE]: ["тридцатому", "тридцатым"],
      [I.ACCUSATIVE]: ["тридцатое", "тридцатые"],
      [I.INSTRUMENTAL]: ["тридцатым", "тридцатыми"],
      [I.PREPOSITIONAL]: ["тридцатом", "тридцатых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["тридцатая", "тридцатые"],
      [I.GENITIVE]: ["тридцатой", "тридцатых"],
      [I.DATIVE]: ["тридцатой", "тридцатым"],
      [I.ACCUSATIVE]: ["тридцатую", "тридцатые"],
      [I.INSTRUMENTAL]: ["тридцатой", "тридцатыми"],
      [I.PREPOSITIONAL]: ["тридцатой", "тридцатых"]
    }
  },
  {
    // 40
    [T.MALE]: {
      [I.NOMINATIVE]: ["сороковой", "сороковые"],
      [I.GENITIVE]: ["сорокового", "сороковых"],
      [I.DATIVE]: ["сороковому", "сороковым"],
      [I.ACCUSATIVE]: ["сороковой", "сороковые"],
      [I.INSTRUMENTAL]: ["сороковым", "сороковыми"],
      [I.PREPOSITIONAL]: ["сороковом", "сороковых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["сороковое", "сороковые"],
      [I.GENITIVE]: ["сорокового", "сороковых"],
      [I.DATIVE]: ["сороковому", "сороковым"],
      [I.ACCUSATIVE]: ["сороковое", "сороковые"],
      [I.INSTRUMENTAL]: ["сороковым", "сороковыми"],
      [I.PREPOSITIONAL]: ["сороковом", "сороковых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["сороковая", "сороковые"],
      [I.GENITIVE]: ["сороковой", "сороковых"],
      [I.DATIVE]: ["сороковой", "сороковым"],
      [I.ACCUSATIVE]: ["сороковую", "сороковые"],
      [I.INSTRUMENTAL]: ["сороковой", "сороковыми"],
      [I.PREPOSITIONAL]: ["сороковой", "сороковых"]
    }
  },
  {
    // 50
    [T.MALE]: {
      [I.NOMINATIVE]: ["пятидесятый", "пятидесятые"],
      [I.GENITIVE]: ["пятидесятого", "пятидесятых"],
      [I.DATIVE]: ["пятидесятому", "пятидесятым"],
      [I.ACCUSATIVE]: ["пятидесятый", "пятидесятые"],
      [I.INSTRUMENTAL]: ["пятидесятым", "пятидесятыми"],
      [I.PREPOSITIONAL]: ["пятидесятом", "пятидесятых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["пятидесятое", "пятидесятые"],
      [I.GENITIVE]: ["пятидесятого", "пятидесятых"],
      [I.DATIVE]: ["пятидесятому", "пятидесятым"],
      [I.ACCUSATIVE]: ["пятидесятое", "пятидесятые"],
      [I.INSTRUMENTAL]: ["пятидесятым", "пятидесятыми"],
      [I.PREPOSITIONAL]: ["пятидесятом", "пятидесятых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["пятидесятая", "пятидесятые"],
      [I.GENITIVE]: ["пятидесятой", "пятидесятых"],
      [I.DATIVE]: ["пятидесятой", "пятидесятым"],
      [I.ACCUSATIVE]: ["пятидесятую", "пятидесятые"],
      [I.INSTRUMENTAL]: ["пятидесятой", "пятидесятыми"],
      [I.PREPOSITIONAL]: ["пятидесятой", "пятидесятых"]
    }
  },
  {
    // 60
    [T.MALE]: {
      [I.NOMINATIVE]: ["шестидесятый", "шестидесятые"],
      [I.GENITIVE]: ["шестидесятого", "шестидесятых"],
      [I.DATIVE]: ["шестидесятому", "шестидесятым"],
      [I.ACCUSATIVE]: ["шестидесятый", "шестидесятые"],
      [I.INSTRUMENTAL]: ["шестидесятым", "шестидесятыми"],
      [I.PREPOSITIONAL]: ["шестидесятом", "шестидесятых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["шестидесятое", "шестидесятые"],
      [I.GENITIVE]: ["шестидесятого", "шестидесятых"],
      [I.DATIVE]: ["шестидесятому	", "шестидесятым"],
      [I.ACCUSATIVE]: ["шестидесятое", "шестидесятые"],
      [I.INSTRUMENTAL]: ["шестидесятым", "шестидесятыми"],
      [I.PREPOSITIONAL]: ["шестидесятом", "шестидесятых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["шестидесятая", "шестидесятые"],
      [I.GENITIVE]: ["шестидесятой", "шестидесятых"],
      [I.DATIVE]: ["шестидесятой", "шестидесятым"],
      [I.ACCUSATIVE]: ["шестидесятую", "шестидесятые"],
      [I.INSTRUMENTAL]: ["шестидесятой", "шестидесятыми"],
      [I.PREPOSITIONAL]: ["шестидесятой", "шестидесятых"]
    }
  },
  {
    // 70
    [T.MALE]: {
      [I.NOMINATIVE]: ["семидесятый", "семидесятые"],
      [I.GENITIVE]: ["семидесятого", "семидесятых"],
      [I.DATIVE]: ["семидесятому", "семидесятым"],
      [I.ACCUSATIVE]: ["семидесятый", "семидесятые"],
      [I.INSTRUMENTAL]: ["семидесятым", "семидесятыми"],
      [I.PREPOSITIONAL]: ["семидесятом", "семидесятых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["семидесятое", "семидесятые"],
      [I.GENITIVE]: ["семидесятого", "семидесятых"],
      [I.DATIVE]: ["семидесятому", "семидесятым"],
      [I.ACCUSATIVE]: ["семидесятое", "семидесятые"],
      [I.INSTRUMENTAL]: ["семидесятым", "семидесятыми"],
      [I.PREPOSITIONAL]: ["семидесятом", "семидесятых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["семидесятая", "семидесятые"],
      [I.GENITIVE]: ["семидесятой", "семидесятых"],
      [I.DATIVE]: ["семидесятой", "семидесятым"],
      [I.ACCUSATIVE]: ["семидесятую", "семидесятые"],
      [I.INSTRUMENTAL]: ["семидесятой", "семидесятыми"],
      [I.PREPOSITIONAL]: ["семидесятой", "семидесятых"]
    }
  },
  {
    // 80
    [T.MALE]: {
      [I.NOMINATIVE]: ["восьмидесятый", "восьмидесятые"],
      [I.GENITIVE]: ["восьмидесятого", "восьмидесятых"],
      [I.DATIVE]: ["восьмидесятому", "восьмидесятым"],
      [I.ACCUSATIVE]: ["восьмидесятый", "восьмидесятые"],
      [I.INSTRUMENTAL]: ["восьмидесятым", "восьмидесятыми"],
      [I.PREPOSITIONAL]: ["восьмидесятом", "восьмидесятых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["восьмидесятое", "восьмидесятые"],
      [I.GENITIVE]: ["восьмидесятого", "восьмидесятых"],
      [I.DATIVE]: ["восьмидесятому", "восьмидесятым"],
      [I.ACCUSATIVE]: ["восьмидесятое", "восьмидесятые"],
      [I.INSTRUMENTAL]: ["восьмидесятым", "восьмидесятыми"],
      [I.PREPOSITIONAL]: ["восьмидесятом", "восьмидесятых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["восьмидесятая", "восьмидесятые"],
      [I.GENITIVE]: ["восьмидесятой", "восьмидесятых"],
      [I.DATIVE]: ["восьмидесятой", "восьмидесятым"],
      [I.ACCUSATIVE]: ["восьмидесятую", "восьмидесятые"],
      [I.INSTRUMENTAL]: ["восьмидесятой", "восьмидесятыми"],
      [I.PREPOSITIONAL]: ["восьмидесятой", "восьмидесятых"]
    }
  },
  {
    // 90
    [T.MALE]: {
      [I.NOMINATIVE]: ["девяностый", "девяностые"],
      [I.GENITIVE]: ["девяностого", "девяностых"],
      [I.DATIVE]: ["девяностому", "девяностым"],
      [I.ACCUSATIVE]: ["девяностый", "девяностые"],
      [I.INSTRUMENTAL]: ["девяностым", "девяностыми"],
      [I.PREPOSITIONAL]: ["девяностом", "девяностых"]
    },
    [T.NEUTER]: {
      [I.NOMINATIVE]: ["девяностое", "девяностые"],
      [I.GENITIVE]: ["девяностого", "девяностых"],
      [I.DATIVE]: ["девяностому", "девяностым"],
      [I.ACCUSATIVE]: ["девяностое", "девяностые"],
      [I.INSTRUMENTAL]: ["девяностым", "девяностыми"],
      [I.PREPOSITIONAL]: ["девяностом", "девяностых"]
    },
    [T.FEMALE]: {
      [I.NOMINATIVE]: ["девяностая", "девяностые"],
      [I.GENITIVE]: ["девяностой", "девяностых"],
      [I.DATIVE]: ["девяностой", "девяностым"],
      [I.ACCUSATIVE]: ["девяностую", "девяностые"],
      [I.INSTRUMENTAL]: ["девяностой", "девяностыми"],
      [I.PREPOSITIONAL]: ["девяностой", "девяностых"]
    }
  }
], G = {
  digits: sI,
  tenToNineteen: oI,
  tens: SI,
  hundreds: cI
}, y = {
  digit: [
    "",
    "одно",
    "двух",
    // "двухтысячная"
    "трёх",
    "четырёх",
    "пяти",
    "шести",
    "семи",
    "восьми",
    "девяти"
  ],
  tenToNineteen: [
    "десяти",
    "одиннадцати",
    "двенадцати",
    // "двенадцатитысячная"
    "тринадцати",
    "четырнадцати",
    "пятнадцати",
    "шестнадцати",
    "семидесяти",
    "восемнадцати",
    "девятнадцати"
  ],
  tens: [
    "",
    "",
    "двадцати",
    // "двадцатитысячная"
    "тридцати",
    "сорока",
    "пятидесяти",
    "шестидесяти",
    "семидесяти",
    "восьмидесяти",
    "девяносто"
  ],
  hundreds: [
    "",
    "сто",
    "двухсот",
    // "двухсоттысячная"
    "трёхсот",
    "четырёхсот",
    "пятисот",
    "шестисот",
    "семисот",
    "восьмисот",
    "девятисот"
  ]
};
function OI(e, A = 1, N = "nominative") {
  if (e.length < 1)
    return "";
  let t = "";
  const E = nI(e), V = [...E].reverse().findIndex((n) => n !== "000"), r = E.length - V - 1;
  if (V === -1) {
    const n = G.digits[0];
    return b(
      n[T.FEMALE],
      N,
      A !== 0
    );
  }
  if (E.length - 1 - V > 0) {
    const n = E.map(
      (s, S) => S === r ? "000" : s
    );
    t += P(
      n,
      1,
      I.NOMINATIVE
    ).result + " ";
  }
  if (V >= 1) {
    const n = E[r].split("").map((M) => parseInt(M)), s = y.hundreds[n[0]], S = n[1] === 1 ? y.tenToNineteen[n[2]] : y.tens[n[1]];
    let c = n[1] === 1 ? "" : y.digit[n[2]];
    rI(n, [0, 0, 1]) && E.length - 1 === V && (c = "");
    const o = V <= h.length ? h[V - 1] : j[V - 2], O = b(
      w,
      N,
      A !== 0
    );
    t += `${s}${S}${c}${o}${O} `;
  }
  if (V === 0) {
    const n = E[r].split("").map((i) => parseInt(i)), s = [...n].reverse().findIndex((i) => i !== 0), S = n.length - s - 1, c = [
      n.map((i, L) => L === S ? 0 : i).map((i, L) => L === 1 && i === 1 ? 0 : i).join("")
    ];
    c[0] !== "000" && (t += P(
      c,
      1,
      I.NOMINATIVE
    ).result + " ");
    let o = n[S], O = G.hundreds;
    s === 0 && (O = G.digits), s === 1 && (O = G.tens), n[1] === 1 && (O = G.tenToNineteen, o = n[2]);
    const M = b(
      O[o][T.FEMALE],
      N,
      A !== 0
    );
    t += `${M} `;
  }
  return t.trim();
}
function iI(e, A) {
  if (e[2] === "/")
    return e;
  const N = [...e], t = A.fractionalPartMinLength || 0;
  return N[3].length < t && (N[3] = N[3] + "0".repeat(t - N[3].length)), N;
}
function a(e, A, N) {
  const t = Array.isArray(A) ? A : A.split(".");
  let E = e;
  for (const V of t) {
    if (E == null || !Object.prototype.hasOwnProperty.call(E, V))
      return N;
    E = E[V];
  }
  return E;
}
const u = (e, A, N) => {
  const t = Array.isArray(A) ? A : A.match(/([^[.\]])+/g);
  t == null || t.reduce((E, V, r) => (E[V] === void 0 && (E[V] = {}), r === t.length - 1 && (E[V] = N), E[V]), e);
};
function MI(e, A) {
  for (const N in e)
    Object.prototype.hasOwnProperty.call(e, N) && A(e[N], N);
}
function aI(e) {
  if (typeof e != "object" || e === null)
    return !1;
  const A = Object.getPrototypeOf(e);
  if (A === null)
    return !0;
  const N = Object.prototype.hasOwnProperty.call(A, "constructor") && A.constructor;
  return typeof N == "function" && N instanceof N && Function.prototype.toString.call(N) === Function.prototype.toString.call(Object);
}
function H(e, A) {
  let N = structuredClone(e);
  const t = (E, V = []) => {
    MI(E, (r, n) => {
      aI(r) ? r = t(r, [...V, n]) : u(N, [...V, n], A(V, n, r));
    });
  };
  return t(e), N;
}
function LI(e, A) {
  return H(e, (N, t, E) => Object.prototype.toString.call(a(A, [...N, t])) === Object.prototype.toString.call(E) ? a(A, [...N, t]) : E);
}
const W = {
  currency: "rub",
  declension: I.NOMINATIVE,
  roundNumber: -1,
  convertMinusSignToWord: !0,
  showNumberParts: {
    integer: !0,
    fractional: !0
  },
  convertNumberToWords: {
    integer: !0,
    fractional: !1
  },
  showCurrency: {
    integer: !0,
    fractional: !0
  }
}, F = {
  fractionalPartMinLength: 2
}, X = {
  number: {
    fractionalPartMinLength: 0,
    currencyNameCases: ["целая", "целых", "целых"],
    currencyNameDeclensions: {
      [I.NOMINATIVE]: ["целая", ""],
      [I.GENITIVE]: ["целой", "целых"],
      [I.DATIVE]: ["целой", "целым"],
      [I.ACCUSATIVE]: ["целую", "целые"],
      [I.INSTRUMENTAL]: ["целой", "целыми"],
      [I.PREPOSITIONAL]: ["целой", "целых"]
    },
    fractionalPartNameCases: ["", "", ""],
    fractionalPartNameDeclensions: {
      [I.NOMINATIVE]: ["", ""],
      [I.GENITIVE]: ["", ""],
      [I.DATIVE]: ["", ""],
      [I.ACCUSATIVE]: ["", ""],
      [I.INSTRUMENTAL]: ["", ""],
      [I.PREPOSITIONAL]: ["", ""]
    },
    currencyNounGender: {
      integer: 1,
      fractionalPart: 1
    }
  },
  rub: {
    ...F,
    currencyNameCases: ["рубль", "рубля", "рублей"],
    currencyNameDeclensions: {
      [I.NOMINATIVE]: ["рубль", ""],
      [I.GENITIVE]: ["рубля", "рублей"],
      [I.DATIVE]: ["рублю", "рублям"],
      [I.ACCUSATIVE]: ["рубль", ""],
      [I.INSTRUMENTAL]: ["рублём", "рублями"],
      [I.PREPOSITIONAL]: ["рубле", "рублях"]
    },
    fractionalPartNameCases: ["копейка", "копейки", "копеек"],
    fractionalPartNameDeclensions: {
      [I.NOMINATIVE]: ["копейка", ""],
      [I.GENITIVE]: ["копейки", "копеек"],
      [I.DATIVE]: ["копейке", "копейкам"],
      [I.ACCUSATIVE]: ["копейку", ""],
      [I.INSTRUMENTAL]: ["копейкой", "копейками"],
      [I.PREPOSITIONAL]: ["копейке", "копейках"]
    },
    currencyNounGender: {
      integer: 0,
      // Мужской род
      fractionalPart: 1
      // Женский род
    }
  },
  usd: {
    ...F,
    currencyNameCases: ["доллар", "доллара", "долларов"],
    currencyNameDeclensions: {
      [I.NOMINATIVE]: ["доллар", ""],
      [I.GENITIVE]: ["доллара", "долларов"],
      [I.DATIVE]: ["доллару", "долларам"],
      [I.ACCUSATIVE]: ["доллар", ""],
      [I.INSTRUMENTAL]: ["долларом", "долларами"],
      [I.PREPOSITIONAL]: ["долларе", "долларах"]
    },
    fractionalPartNameCases: ["цент", "цента", "центов"],
    fractionalPartNameDeclensions: {
      [I.NOMINATIVE]: ["цент", ""],
      [I.GENITIVE]: ["цента", "центов"],
      [I.DATIVE]: ["центу", "центам"],
      [I.ACCUSATIVE]: ["цент", ""],
      [I.INSTRUMENTAL]: ["центом", "центами"],
      [I.PREPOSITIONAL]: ["центе", "центах"]
    },
    currencyNounGender: {
      integer: 0,
      fractionalPart: 0
    }
  },
  eur: {
    ...F,
    currencyNameCases: ["евро", "евро", "евро"],
    currencyNameDeclensions: {
      [I.NOMINATIVE]: ["евро", ""],
      [I.GENITIVE]: ["евро", "евро"],
      [I.DATIVE]: ["евро", "евро"],
      [I.ACCUSATIVE]: ["евро", ""],
      [I.INSTRUMENTAL]: ["евро", "евро"],
      [I.PREPOSITIONAL]: ["евро", "евро"]
    },
    fractionalPartNameCases: ["цент", "цента", "центов"],
    fractionalPartNameDeclensions: {
      [I.NOMINATIVE]: ["цент", ""],
      [I.GENITIVE]: ["цента", "центов"],
      [I.DATIVE]: ["центу", "центам"],
      [I.ACCUSATIVE]: ["цент", ""],
      [I.INSTRUMENTAL]: ["центом", "центами"],
      [I.PREPOSITIONAL]: ["центе", "центах"]
    },
    currencyNounGender: {
      integer: 0,
      fractionalPart: 0
    }
  }
};
function RI(e) {
  const A = structuredClone(e.currency), N = W.currency, t = X[N];
  let E = structuredClone(
    t
  );
  return E = LI(
    E,
    A
  ), a(A, [
    "currencyNameDeclensions",
    I.NOMINATIVE,
    0
  ]) === void 0 && a(A, ["currencyNameCases", 0]) !== void 0 && u(
    E,
    ["currencyNameDeclensions", I.NOMINATIVE, 0],
    A.currencyNameCases[0]
  ), a(A, [
    "currencyNameDeclensions",
    I.GENITIVE,
    0
  ]) === void 0 && a(A, ["currencyNameCases", 1]) !== void 0 && u(
    E,
    ["currencyNameDeclensions", I.GENITIVE, 0],
    A.currencyNameCases[1]
  ), a(A, [
    "currencyNameDeclensions",
    I.GENITIVE,
    1
  ]) === void 0 && a(A, ["currencyNameCases", 2]) !== void 0 && u(
    E,
    ["currencyNameDeclensions", I.GENITIVE, 1],
    A.currencyNameCases[2]
  ), a(A, [
    "fractionalPartNameDeclensions",
    I.NOMINATIVE,
    0
  ]) === void 0 && a(A, ["fractionalPartNameCases", 0]) !== void 0 && u(
    E,
    ["fractionalPartNameDeclensions", I.NOMINATIVE, 0],
    A.fractionalPartNameCases[0]
  ), a(A, [
    "fractionalPartNameDeclensions",
    I.GENITIVE,
    0
  ]) === void 0 && a(A, ["fractionalPartNameCases", 1]) !== void 0 && u(
    E,
    ["fractionalPartNameDeclensions", I.GENITIVE, 0],
    A.fractionalPartNameCases[1]
  ), a(A, [
    "fractionalPartNameDeclensions",
    I.GENITIVE,
    1
  ]) === void 0 && a(A, ["fractionalPartNameCases", 2]) !== void 0 && u(
    E,
    ["fractionalPartNameDeclensions", I.GENITIVE, 1],
    A.fractionalPartNameCases[2]
  ), delete E.currencyNameCases, delete E.fractionalPartNameCases, E;
}
function uI(e = {}) {
  let A = {};
  const N = (e == null ? void 0 : e.currency) || W.currency;
  return typeof N == "string" && (A = X[N]), typeof N == "object" && (A = RI(e)), A;
}
function J(e, A, N, t, E, V) {
  var s, S, c, o;
  const r = A === "integer" ? e.currencyNameDeclensions : e.fractionalPartNameDeclensions;
  let n = (s = r == null ? void 0 : r[V]) == null ? void 0 : s[N === 0 ? 0 : 1];
  return (V === I.NOMINATIVE || V === I.ACCUSATIVE) && N >= 1 && (n = (S = r == null ? void 0 : r[I.GENITIVE]) == null ? void 0 : S[N === 1 ? 0 : 1], E === "number" && (n = (c = r == null ? void 0 : r[I.GENITIVE]) == null ? void 0 : c[1])), t === !0 && (n = (o = r == null ? void 0 : r[I.GENITIVE]) == null ? void 0 : o[1]), n || "";
}
function CI(e, A, N = "nominative", t = 0) {
  var r;
  e < 0 && (e = 0);
  let E = "", V = {};
  if (e <= z.length - 1)
    V = z[e];
  else {
    const n = Math.ceil((e + 2) / 3) - 1, s = e - n * 3 + 1, S = n <= h.length ? h[n - 1] : j[n - 2], c = VI[s];
    Object.keys(w).forEach((o) => {
      w[o].forEach((M, i) => {
        u(
          V,
          [o, i],
          `${c}${S}${M}`
        );
      });
    });
  }
  return E = b(
    V,
    N,
    t !== 0
  ), A === 0 && (E = (r = V[I.GENITIVE]) == null ? void 0 : r[1]), E;
}
function UI(e, A) {
  const N = Array.isArray(A) ? A : A.split(".");
  let t = e;
  for (const E of N) {
    if (t == null || !Object.prototype.hasOwnProperty.call(t, E))
      return !1;
    t = t[E];
  }
  return !0;
}
function PI(e = {}) {
  let A = structuredClone(e);
  return A.convertNumberToWords || A.convertNumbertToWords && (A.convertNumberToWords = A.convertNumbertToWords), delete A.convertNumbertToWords, A;
}
function Z(e = {}) {
  const A = PI(e);
  return H(
    W,
    (t, E, V) => UI(A, [...t, E]) ? a(A, [...t, E]) : V
  );
}
function g(e) {
  const A = e.length, E = Math.ceil(A / 3) * 3 - A, V = "0".repeat(E) + e, r = [];
  for (let n = 0; n < V.length; n += 3) {
    const s = V[n] + V[n + 1] + V[n + 2];
    r.push(s);
  }
  return r;
}
function Y(e, A, N) {
  return e.substring(0, A) + N.toString() + e.substring(A + N.toString().length);
}
function k(e, A = 2) {
  if (e[2] === "/")
    return e;
  let N = [...e];
  if (A >= 0) {
    if (e[3].length <= A)
      return e;
    const t = e[1], E = e[3].substring(0, A + 1);
    let V = `${t}.${E}`, r = !1;
    for (let n = V.length - 1; n >= 0; n--)
      if (V[n].search(/[0-9]/) !== -1) {
        const s = parseInt(V[n]);
        if (r === !0)
          if (s === 9)
            V = Y(V, n, 0), r = !0, n === 0 && (V = `1${V}`);
          else {
            V = Y(
              V,
              n,
              s + 1
            ), r = !1;
            break;
          }
        else {
          if (s <= 4)
            break;
          r = !0;
        }
      }
    N[1] = V.slice(0, -1).split(".")[0], N[3] = V.slice(0, -1).split(".")[1];
  }
  return N[3] = N[3].split("").reverse().join("").replace(/^0+/, "").split("").reverse().join(""), N[3] === "" && A !== 0 && (N[3] = "0"), N;
}
function lI(e, A) {
  var c, o, O, M, i, L, m, d, _, v;
  const N = ["", "", "", "", ""];
  let t = [...e];
  const E = Z(A), V = uI(E);
  e[0] === "-" && (E.convertMinusSignToWord === !0 ? N[0] = tI : N[0] = "-"), t = k(e, E.roundNumber), typeof E.currency == "string" && E.currency !== "number" && (t = k(t, 2)), t = iI(
    t,
    V
  );
  const r = t[1], n = t[3], s = t[2];
  if (((c = E.showNumberParts) == null ? void 0 : c.integer) === !0) {
    N[1] = r;
    const R = P(
      g(r),
      (o = V.currencyNounGender) == null ? void 0 : o.integer,
      E.declension
    );
    if (((O = E.convertNumberToWords) == null ? void 0 : O.integer) === !0 && (s !== "/" ? N[1] = R.result : N[1] = P(
      g(r),
      1,
      E.declension
    ).result), ((M = E.showCurrency) == null ? void 0 : M.integer) === !0 && s !== "/") {
      const C = J(
        V,
        "integer",
        R.unitNameForm,
        R.lastScaleIsZero,
        // @ts-expect-error
        E.currency,
        E.declension
      );
      N[2] = C;
    }
  }
  if (((i = E.showNumberParts) == null ? void 0 : i.fractional) === !0) {
    N[3] = n;
    const R = P(
      g(n),
      (L = V.currencyNounGender) == null ? void 0 : L.fractionalPart,
      E.declension
    );
    if (((m = E.convertNumberToWords) == null ? void 0 : m.fractional) === !0)
      if (s === "/") {
        const C = P(
          g(r),
          (d = V.currencyNounGender) == null ? void 0 : d.integer,
          E.declension
        );
        N[3] = OI(
          g(n),
          C.unitNameForm,
          E.declension
        );
      } else
        N[3] = R.result;
    if (((_ = E.convertNumberToWords) == null ? void 0 : _.fractional) === !1 && E.currency === "number" && N[3].length > 0 && (N[3] = N[3].replace(/^0+/, ""), N[3] === "" && E.roundNumber !== 0 && (N[3] = "0")), ((v = E.showCurrency) == null ? void 0 : v.fractional) === !0) {
      if (E.currency !== "number") {
        const C = J(
          V,
          "fractional",
          R.unitNameForm,
          R.lastScaleIsZero,
          // @ts-expect-error
          E.currency,
          E.declension
        );
        N[3] !== "" && (N[4] = C);
      }
      if (E.currency === "number" && s !== "/" && // @ts-expect-error
      (E.roundNumber > 0 || // @ts-expect-error
      E.roundNumber < 0 && n.length > 0)) {
        const C = parseInt(
          n[n.length - 1]
        );
        N[4] = CI(
          n.length - 1,
          C,
          E.declension,
          R.unitNameForm
        );
      }
      s === "/" && E.currency !== "number" && (N[4] = // @ts-expect-error
      V.currencyNameDeclensions[I.GENITIVE][0]);
    }
  }
  let S = N.filter(Boolean).join(" ").replace(/\s+/g, " ").trim();
  return S = S.charAt(0).toUpperCase() + S.slice(1), S;
}
function fI(e) {
  const N = typeof e == "string" || typeof e == "number" ? e.toString() : "0";
  let t = [];
  t[2] = ".";
  let E = N.replace(/[^\d\.\,\/\-]/g, "");
  return E = E.length < 1 ? "0" : E, t[0] = E.search(/\-/) === 0 ? "-" : "+", E = E.replace(/[\-]/g, ""), E.search(/[\,\.\/]/) > -1 && (t[2] = E.substr(E.search(/[\,\.\/]/), 1)), E = E.replace(/[\,\.\/]/, "|CUTHERE|"), E = E.replace(/[\,\.\/]/g, ""), t[1] = E.split("|CUTHERE|")[0], t[3] = E.split("|CUTHERE|")[1], t[3] = t[3] === void 0 ? "" : t[3], t[1] = t[1].replace(/^0+/, ""), t[2] !== "/" && (t[3] = t[3].split("").reverse().join("").replace(/^0+/, "").split("").reverse().join("")), t[1] = t[1] === "" ? "0" : t[1], t[3] = t[3] === "" ? "0" : t[3], t[1].length > 306 && (t[1] = t[1].slice(0, 306)), t[3].length > 306 - 1 && (t[3] = t[3].slice(0, 306 - 1)), t;
}
function q(e, A) {
  const N = Z(A), t = fI(e);
  return lI(t, N);
}
"structuredClone" in globalThis || (globalThis.structuredClone = TI);
const DI = {
  convert: q
}, GI = q;
export {
  GI as convert,
  DI as default
};
